Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,656 @@
+      hwpart->source = fake_mac;
+
+      *ippart = Common::IPv4Header(20, IPPROTO_TCP, network_ref[i].from, network_ref[i].to);
+
+      *tcppart = Common::TCPHeader(network_ref[i].from, network_ref[i].to, network_ref[i].seq_num,
+                                   network_ref[i].ack_num, TCP_FLAG_ACK);
+      tcppart->checksum = Common::ComputeTCPNetworkChecksum(network_ref[i].from, network_ref[i].to,
+                                                            tcppart, 20, IPPROTO_TCP);
+
+      WriteToQueue(m_in_frame, 0x36);
+    }
+
+    // clear any ack data
+    if (tcpdata->properties & TCP_FLAG_ACK)
+    {
+      for (auto& tcp_buf : network_ref[i].TcpBuffers)
+      {
+        if (tcp_buf.used && tcp_buf.seq_id < Common::swap32(tcpdata->acknowledgement_number))
+        {
+          tcp_buf.used = false;  // confirmed data received
+          if (!network_ref[i].ready && !network_ref[i].TcpBuffers[0].used)
+            network_ref[i].ready = true;
+        }
+      }
+    }
+  }
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::HandleUDPFrame(Common::EthernetHeader* hwdata,
+                                                       Common::IPv4Header* ipdata,
+                                                       Common::UDPHeader* udpdata, u8* data)
+{
+  sf::IpAddress target;
+
+  if (*(u32*)ipdata->destination_addr == 0)
+    *(u32*)ipdata->destination_addr = 0x0101a8c0;
+  // dns request
+  int i = GetAvaibleSlot(udpdata->destination_port);
+  if (network_ref[i].ip == 0)
+  {
+    network_ref[i].ip = *(u32*)ipdata->destination_addr;  // change for ip
+    network_ref[i].local = udpdata->source_port;
+    network_ref[i].remote = udpdata->destination_port;
+    network_ref[i].type = 17;
+    memcpy(&network_ref[i].bba_mac, &m_eth_ref->mBbaMem[BBA_NAFR_PAR0], Common::MAC_ADDRESS_SIZE);
+    network_ref[i].my_mac = fake_mac;
+    network_ref[i].from.sin_addr.s_addr = *(u32*)&ipdata->destination_addr;
+    network_ref[i].from.sin_port = udpdata->destination_port;
+    network_ref[i].to.sin_addr.s_addr = *(u32*)&ipdata->source_addr;
+    network_ref[i].to.sin_port = udpdata->source_port;
+    network_ref[i].udp_socket.setBlocking(false);
+    if (network_ref[i].udp_socket.bind(sf::Socket::AnyPort) != sf::Socket::Done)
+    {
+      ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+      return;
+    }
+  }
+  if (Common::swap16(udpdata->destination_port) == 53)
+  {
+    target = sf::IpAddress(m_dns_ip.c_str());  // dns server ip
+  }
+  else
+  {
+    target = sf::IpAddress(Common::swap32(*(u32*)ipdata->destination_addr));
+  }
+
+  network_ref[i].udp_socket.send(data, Common::swap16(udpdata->length) - 8, target,
+                                 Common::swap16(udpdata->destination_port));
+}
+
+bool CEXIETHERNET::BuiltInBBAInterface::SendFrame(const u8* frame, u32 size)
+{
+  Common::EthernetHeader* hwdata;
+  Common::IPv4Header* ipdata;
+  Common::UDPHeader* udpdata;
+  Common::TCPHeader* tcpdata;
+
+  int offset = 0;
+  mtx.lock();
```

Comment:
Ditto for the [lock guard](https://en.cppreference.com/w/cpp/thread/lock_guard).